src/utils/async.js

```diff
@@ -0,0 +1,245 @@
+/* eslint-disable max-classes-per-file */
+import { MapWithCapacity } from './generic';
+
+/**
+ * A promise that can be reused.
+ */
+export class ReusablePromise {
+  /**
+   *
+   * @param {object} configuration
+   * @param {boolean} [configuration.save=false] - if set to `true`, stores the result of the first successful promise forever.
+   */
+  constructor({ save = false } = {}) {
+    this.promise = null;
+    this.save = save;
+  }
+
+  /**
+   * Checks if a promise exists, if so, `await`s it, otherwise inserts a new `Promise` produced by
+   * calling `createPromise` function.
+   *
+   * @template T
+   * @param {function(): Promise<T>} createPromise
+   * @returns {Promise<T>}
+   */
+  async call(createPromise) {
+    if (this.promise != null) {
+      return await this.promise;
+    }
+    const promise = createPromise();
+    this.promise = promise;
+
+    let res;
+    let err;
+    try {
+      res = await this.promise;
+    } catch (e) {
+      err = e;
+    } finally {
+      if ((!this.save || err != null) && this.promise === promise) {
+        this.clear();
+      }
+    }
+
+    if (err != null) {
+      throw err;
+    }
+    return res;
+  }
+
+  clear() {
+    this.promise = null;
+  }
+}
+
+/**
+ * A map where each entry represents a keyed promise that can be reused.
+ */
+export class ReusablePromiseMap {
```

Comment:
They're reusable but `ReusablePromiseMap` is particularly needed when you have a keyed request that might be reused and must be cleaned on error(s) [or if `save` was provided].
The `ReusablePromise` and `ReusablePromiseMap` differ from `Promise` even with `save=true` because they clear the underlying value if an error is encountered.
If `save=false`, they have completely different behavior because they will be cleared whenever the promise is fulfilled.

Reply:
Sorry, i meant this comment on `ReusablePromise`. So in `BlocksProvider.lastHash`, if you had 
```
this.lastHashCall = this.finalized
      ? this.api.rpc.chain.getFinalizedHead()
      : this.api.rpc.chain.getBlockHash()
```

rather than
```
this.lastHashCall.call(() => (this.finalized
      ? this.api.rpc.chain.getFinalizedHead()
      : this.api.rpc.chain.getBlockHash()))
```

Reply:
Then we would have an issue if an error was encountered. Also, we do need to keep a value of the promise only while it’s not resolved, and clean it once it’s fulfilled.

Reply:
I see.